src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,201 @@
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_certificateFile == null)
+                {
+                    return;
+                }
+                try
+                {
+                    File.Delete(_certificateFile);
+                }
+                catch (IOException e)
+                {
+                    VerboseTrace("Deleting temporary file:" + e);
+                }
+            }
+
+            internal static ClientCertProvider Create(ClientCertificateOption option, string uriScheme)
+            {
+                if (option == ClientCertificateOption.Manual || !string.Equals(uriScheme, UriSchemeHttps))
+                {
+                    return null;
+                }
+
+                using (var myStore = new X509Store())
+                {
+                    myStore.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection clientCerts = myStore.Certificates.FindTlsClientCertificates();
+                    if (clientCerts.Count == 0)
+                    {
+                        VerboseTrace("0 Client Certs");
+                        return null; 
+                    }
+                    VerboseTrace("Client Certs found " + clientCerts.Count);
+
+                    // NOTE: This is a DEVIATION FROM WinHttpHandler
+                    // Unlike WinHttpHandler, we are going to use only one client certificate.
+                    // libcurl, libssl and even X509Certificate2Collection.Export all are able to handle single client certificate.
+                    // Picking the first available cert
+                    return new ClientCertProvider(clientCerts[0]);
+                }
+            }
+
+            private static int CurlSslCallback(IntPtr easyHandle, IntPtr sslContext, IntPtr x509Handle)
+            {
+                Debug.Assert(x509Handle != null);
+                Debug.Assert(sslContext != null);
+
+                Interop.libcurl.SSL_CTX_use_certificate(sslContext, x509Handle);
+
+                // TODO: Need to set the private key. Fix when dev_sslstream gets merged.
+                // For now setting it to nullPtr
+                Interop.libcurl.SSL_CTX_use_PrivateKey(sslContext, IntPtr.Zero);
+
+                int retCode = Interop.libcurl.SSL_CTX_check_private_key(sslContext);
+                if (retCode != 1)
+                {
+                    // In the current case, this will always throw.
+                    VerboseTrace("SET Cert failed with " + retCode);
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+
+                VerboseTrace("successfully set client certificate");
+                return CURLcode.CURLE_OK;
+            }
+
+            private static string GeneratePassword()
+            {
+                byte[] bytes = new byte[SecurePasswordByteLength];
+                var prng = new Random();
+                prng.NextBytes(bytes);
```

Comment:
> I know you suggested using Random here. Why is it sufficient?

I had to dig around a bit to find what I'd said, because I hoped I hadn't said that...  it turns out to have been:
> The PRNG is the type named Random. RandomNumberGenerator is the cryptographically secure random number generator. rng is good enough, or csrng if you feel fancy :).

My comment wasn't "switch to `Random`", it was "this variable is named incorrectly".  But I can see how that would have come across wrong.

@kapilash Please switch back to using RandomNumberGenerator.Create(), but update the variable name to not be `prng`.